lib/async_hooks.js

```diff
@@ -190,6 +191,123 @@ class AsyncResource {
 }
 
 
+// AsyncLocal //
+
+const kStack = Symbol('stack');
+const kIsFirst = Symbol('is-first');
+const kMap = Symbol('map');
+const kOnChangedCb = Symbol('on-changed-cb');
+const kHooks = Symbol('hooks');
+const kSet = Symbol('set');
+
+class AsyncLocal {
+  constructor(options = {}) {
+    if (typeof options !== 'object' || options === null)
+      throw new ERR_INVALID_ARG_TYPE('options', 'Object', options);
+
+    const { onChangedCb = null } = options;
+    if (onChangedCb !== null && typeof onChangedCb !== 'function')
+      throw new ERR_INVALID_ARG_TYPE('options.onChangedCb',
+                                     'function',
+                                     onChangedCb);
+
+    this[kOnChangedCb] = onChangedCb;
+    this[kMap] = new Map();
+
+    const fns = {
+      init: (asyncId, type, triggerAsyncId, resource) => {
+        // Propagate value from current id to new (execution graph)
+        const value = this[kMap].get(executionAsyncId());
```

Comment:
What about `triggerAsyncId` -- is it ever necessary to use that instead of `executionAsyncId`?

Reply:
Well I thought about it and took a look into implementation of some APMs, Domains and CLS implementation using async hooks. In general the execution path is what is used. I found no place where I had the impression that triggerId was used for a well defined reason, mostly as fallback if there is nothing to propagate from execution context. triggerId will not propagate anything across `await` so using it always is most likely not what anyone wants.
As mentioned above I used .NET as template and there is also no trigger path, just the execution path. Usually a new WebRequest is seen as new transaction so using execution Id seems to me also the better choice.

I thought about adding a configuration option to `AsyncLocal` for a propagation algorithm. En enum like `ExecutionPath`, `TriggerPath` would be easy but full flexibility can be only reached by allowing to install a callback which implements the logic. This comes with the cost to expose async hooks details like resource names, execution/trigger ids which would require an API change once we move to a resource based system instead id based therefore I decided against this for  now. It could be added later if really needed.

Do you have use cases where triggerId is actually what is needed?